prompts/prompt_template_test.go

```diff
@@ -0,0 +1,412 @@
+	t.Parallel()
+
+	err := checkInputVariables([]string{"stop"})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, ErrInputVariableReserved)
+
+	err = checkInputVariables([]string{"test"})
+	require.NoError(t, err)
+}
+
+func TestCheckPartialVariables(t *testing.T) {
+	t.Parallel()
+
+	err := checkPartialVariables(map[string]interface{}{"stop": "test"})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, ErrPartialVariableReserved)
+
+	err = checkPartialVariables(map[string]interface{}{"test": "test"})
+	require.NoError(t, err)
+}
+
+var _ schema.OutputParser[any] = testEmptyOutputParser{}
+
+type testEmptyOutputParser struct{}
+
+func (t testEmptyOutputParser) Parse(_ string) (any, error) {
+	return nil, nil
+}
+
+func (t testEmptyOutputParser) GetFormatInstructions() string {
+	return ""
+}
+
+func (t testEmptyOutputParser) Type() string {
+	return ""
+}
+
+func TestApplyPromptTemplateBaseOptions(t *testing.T) {
+	t.Parallel()
+
+	_, err := applyPromptTemplateBaseOptions(PromptTemplateBaseOptions{
+		InputVariables: []string{"stop"},
+	})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, ErrInputVariableReserved)
+
+	_, err = applyPromptTemplateBaseOptions(PromptTemplateBaseOptions{
+		PartialVariables: map[string]any{"stop": "test"},
+	})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, ErrPartialVariableReserved)
+
+	opts, err := applyPromptTemplateBaseOptions(PromptTemplateBaseOptions{
+		InputVariables:   []string{"test"},
+		OutputParser:     testEmptyOutputParser{},
+		PartialVariables: map[string]any{"test": "test"},
+	})
+	require.NoError(t, err)
+	assert.Equal(t, PromptTemplateBaseOptions{
+		InputVariables:   []string{"test"},
+		OutputParser:     testEmptyOutputParser{},
+		PartialVariables: map[string]any{"test": "test"},
+	}, opts)
+}
+
+func TestFormatPromptValue(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Error", func(t *testing.T) {
+		t.Parallel()
+
+		pt, err := NewPromptTemplate(PromptTemplateOptions{
+			PromptTemplateBaseOptions: PromptTemplateBaseOptions{
+				InputVariables: []string{"testFormatPromptValue"},
+			},
+			Template:         "Hello, {{ .test }}",
+			ValidateTemplate: true,
+		})
+		require.NoError(t, err)
+
+		pt.opts.Template = "Hello, {{{ .testFormatPromptValue }}"
+		_, err = formatPromptValue(pt, map[string]any{
+			"testFormatPromptValue": "world",
+		})
+		require.Error(t, err)
+	})
+
+	t.Run("Success", func(t *testing.T) {
+		t.Parallel()
+
+		pt, err := NewPromptTemplate(PromptTemplateOptions{
+			PromptTemplateBaseOptions: PromptTemplateBaseOptions{
+				InputVariables: []string{"testFormatPromptValue"},
+			},
+			Template:         "Hello, {{ .testFormatPromptValue }}",
+			ValidateTemplate: true,
+		})
+		require.NoError(t, err)
+
+		spv, err := formatPromptValue(pt, map[string]any{
+			"testFormatPromptValue": "world",
+		})
+		require.NoError(t, err)
+		assert.Equal(t, "Hello, world", spv.String())
+	})
+}
+
+func TestMergePartialAndUserVariables(t *testing.T) {
```

Comment:
Seems like a candidate for a table based test

Reply:
Switched.